toolchain/check/import_ref.cpp

```diff
@@ -676,6 +676,74 @@ static auto AddImportIRInst(ImportContext& context, SemIR::InstId inst_id)
       {.ir_id = context.import_ir_id(), .inst_id = inst_id});
 }
 
+// Sets the constant value for an instruction and returns it. The constant value
+// is computed if not explicitly specified.
+static auto ComputeOrSetConstantValue(
+    Context& context, SemIR::InstId inst_id, SemIR::Inst inst,
+    std::optional<SemIR::ConstantId> const_id = std::nullopt)
+    -> SemIR::ConstantId {
+  if (!const_id) {
+    const_id = TryEvalInstUnsafe(context, inst_id, inst);
+  }
+  if (const_id->has_value()) {
+    CARBON_VLOG_TO(context.vlog_stream(), "Constant: {0} -> {1}\n", inst,
+                   context.constant_values().GetInstId(*const_id));
+  }
+  context.constant_values().Set(inst_id, *const_id);
+  return *const_id;
+}
+
+// Adds an imported instruction. The constant value of the instruction can be
+// specified if it is known, otherwise it will be computed.
+//
+// Normally `AddImportedConstant` should be used to create the result of
+// resolving a constant, for example by calling `ResolveAs*`. However, we
+// sometimes need to create a new instruction, not just obtain a constant value,
+// such as when importing a declaration or a pattern. In that case, this
+// function should be used to create it.
+//
+// Do not use `AddInst*`, as it will add symbolic constants to the eval block of
+// whatever generic the import happens within.
+template <typename InstT>
+static auto AddImportedInst(
+    ImportContext& context, SemIR::InstId import_inst_id, InstT inst,
+    std::optional<SemIR::ConstantId> const_id = std::nullopt) -> SemIR::InstId {
+  auto inst_id = context.local_insts().AddInNoBlock(MakeImportedLocIdAndInst(
+      context.local_context(), AddImportIRInst(context, import_inst_id), inst));
+  CARBON_VLOG_TO(context.local_context().vlog_stream(),
+                 "AddImportedInst: {0}\n", static_cast<SemIR::Inst>(inst));
+  ComputeOrSetConstantValue(context.local_context(), inst_id, inst, const_id);
+  return inst_id;
+}
+
+// Convenience for AddImportedInst for placeholder instructions. The constant
+// value is set to `None`; the instruction should later be updated by
+// `ReplaceImoprtedInstAndSetConstantValue`.
+template <typename InstT>
+static auto AddPlaceholderImportedInst(ImportContext& context,
+                                       SemIR::InstId import_decl_id, InstT inst)
+    -> SemIR::InstId {
+  return AddImportedInst(context, import_decl_id, inst,
+                         SemIR::ConstantId::None);
+}
+
+// Replace an imported instruction with a new instruction. The constant value of
+// the instruction must previously have been set to `None`, for example by using
+// `AddPlaceholderImportedInst`. The new constant value can be specified if
```

Comment:
You note "for example": are there other ways import code will do that? I think we should be specific if possible. One answer might be to drop "for example", another might be to add a second example.

IIUC, the `GetLocalConstantIdChecked` call in `SymbolicBindingPattern` code won't ever return `None`, but this comment has me wondering if I'm missing something.

Reply:
The special case for `SymbolicBindingPattern` appears to be unnecessary. Removed that. Also made this more specific.